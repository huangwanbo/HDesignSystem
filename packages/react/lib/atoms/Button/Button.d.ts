import React from "react";
import { buttonSizes, buttonTypes, htmlType } from "../../foundation/ButtonProps";
interface BaseButtonProps {
    children: string | React.ReactNode;
    type?: keyof typeof buttonTypes;
    size?: keyof typeof buttonSizes;
    disabled?: boolean;
    ghost?: boolean;
    onClick?: React.MouseEventHandler<HTMLElement>;
}
declare type AnchorButtonProps = {
    href?: string;
    onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps & Omit<React.AnchorHTMLAttributes<any>, 'type' | 'onClick'>;
declare type NativeButtonProps = {
    htmlType?: keyof typeof htmlType;
    onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps & Omit<React.ButtonHTMLAttributes<any>, 'type' | 'onClick'>;
declare type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>;
declare const Button: React.FunctionComponent<ButtonProps>;
export default Button;
