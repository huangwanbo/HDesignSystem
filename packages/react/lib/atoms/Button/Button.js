import React from 'react';
import cls from 'classNames';

const prefix = "ds-button__container";
const Button = React.forwardRef((props, ref) => {
    const { children, type = "primary", size = "base", disabled = false, ghost = false, href, onClick, htmlType = 'button', ...rest } = props;
    let component = 'button';
    const buttonRef = ref || React.createRef();
    if (href) {
        component = 'a';
    }
    const createOtherProps = () => {
        const otherProps = {};
        if (href) {
            otherProps.href = href;
        }
        return otherProps;
    };
    const cls$1 = cls(prefix, {
        [`${prefix}-${type}`]: !disabled,
        [`${prefix}-${size}`]: true,
        [`${prefix}-${type}-ghost`]: ghost,
        [`${prefix}-${type}-disabled`]: disabled
    });
    const handleClick = (e) => {
        e.preventDefault();
        if (!disabled) {
            onClick?.(e);
        }
    };
    return (React.createElement(component, {
        type: htmlType,
        className: cls$1,
        ref: buttonRef,
        onClick: handleClick,
        ...createOtherProps(),
        ...rest
    }, children));
});
Button.displayName = "Button";

export { Button as default };
//# sourceMappingURL=Button.js.map
