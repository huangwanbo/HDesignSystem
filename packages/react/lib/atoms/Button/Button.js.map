{"version":3,"file":"Button.js","sources":[],"sourcesContent":[],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;"}