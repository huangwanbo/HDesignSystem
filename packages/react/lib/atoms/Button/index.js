export { default } from './Button.js';
//# sourceMappingURL=index.js.map
