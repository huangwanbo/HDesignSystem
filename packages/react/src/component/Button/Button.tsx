import React from "react";
import {
  buttonSizes,
  buttonTypes,
  htmlType,
} from "../../foundation/ButtonProps";
import classNames from "classNames";
const prefix = "ds-button__container";
interface BaseButtonProps {
  children: string | React.ReactNode;
  type?: keyof typeof buttonTypes;
  size?: keyof typeof buttonSizes;
  disabled?: boolean;
  ghost?: boolean;
  onClick?: React.MouseEventHandler<HTMLElement>;
}
type AnchorButtonProps = {
  href?: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps &
  Omit<React.AnchorHTMLAttributes<any>, "type" | "onClick">;
type NativeButtonProps = {
  htmlType?: keyof typeof htmlType;
  onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps &
  Omit<React.ButtonHTMLAttributes<any>, "type" | "onClick">;
export type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>;
const Button: React.FunctionComponent<ButtonProps> = React.forwardRef(
  (props, ref) => {
    const {
      children,
      type = "primary",
      size = "base",
      disabled = false,
      ghost = false,
      href,
      onClick,
      htmlType = "button",
      ...rest
    } = props;
    let component = "button";
    const buttonRef = (ref as any) || React.createRef<ButtonProps>();
    if (href) {
      component = "a";
    }
    const createOtherProps = () => {
      const otherProps: Partial<ButtonProps> = {};
      if (href) {
        otherProps.href = href;
      }

      return otherProps;
    };
    const cls = classNames(prefix, {
      [`${prefix}-${type}`]: !disabled,
      [`${prefix}-${size}`]: true,
      [`${prefix}-${type}-ghost`]: ghost,
      [`${prefix}-${type}-disabled`]: disabled,
    });
    const handleClick = (
      e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>
    ) => {
      e.preventDefault();
      if (!disabled) {
        (
          onClick as React.MouseEventHandler<
            HTMLButtonElement | HTMLAnchorElement
          >
        )?.(e);
      }
    };
    return React.createElement(
      component,
      {
        type: htmlType,
        className: cls,
        ref: buttonRef,
        onClick: handleClick,
        ...createOtherProps(),
        ...rest,
      },
      children
    );
  }
);
Button.displayName = "Button";
export default Button;
