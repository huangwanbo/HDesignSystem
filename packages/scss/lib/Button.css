/* ======================================
    #Variables
====================================== */
/**
* TABLE OF CONTENTS
*
* COLORS
* Brand colors .................... These ar the main brand colors. They will be used the most
* Neutral colors .................. White and black, gray scale, secondary colors.
* Utility colors .................. Other colors to represent states: info, warning, success, etc.
*
* TYPOGRAPHY
* Font families ................... The fonts used in the design system.
* Font weights .................... Font weights
* Sizes ........................... Font sizing for body text, paragraphs, inputs
* Heading sizes ................... Font sizing for page headings
*
* BREAKPOINTS
* Breakpoints .................... Global responsive breakpoints definitions.
*      
* SPACING
* Spacing ......................... Global fixed set of spacing variables.
**/
/*=============================================
=                   colors                    =
=============================================*/
/* Brand Colors */
/* Neutral Colors */
/* Utility colors */
/*=====  End of Color  ======*/
/*=============================================
=            Typography                       =
=============================================*/
/**
 * 1) The playfair font. We'll use the for headings mostly.
 *      https://fonts.google.com/specimen/Playfair+Display
 * 
 * 2) The raleway font. We'll use ths for the body text.
 *      https://fonts.goole.com/specimen/Raleway
*/
/* Font families */
/* 1 */
/* 2 */
/* Font sizes */
/* Heading font sizes */
/* Font weights */
/*=====  End of Typography  ======*/
/*=============================================
=            Breakpoints                      =
=============================================*/
/*=====  End of Breakpoints  ======*/
/*=============================================
=            Spacing                          =
=============================================*/
/*=============================================
=            IconSize                         =
=============================================*/
/*=====  End of IconSize               ======*/
/*=============================================
=            border-radius                    =
=============================================*/
/*=====  End of border-radius          ======*/
/*=============================================
=            Foundation - colors              =
=============================================*/
/**
 * This file defines the actual colors that will be used for styling. They will default to the
 * we defined in the _variable.scss file. This is our default palette, and devs can override this
 * with their own variables.
 */
/*=============================================
 =            Global text colors               =
 =============================================*/
/*=============================================
=                    Buttons                  =
=============================================*/
/*=============================================
=                  Forms                      =
=============================================*/
/*=============================================
=            Minxins                         =
=============================================*/
/**
 * 
 * helpful reusable function-like things that will be used to implement the
 * styles for components.
 * 
*/
.body {
  font-size: 12px;
}

@media (min-width: 48rem) {
  .body {
    font-size: 14px;
  }
}

@media (min-width: 64rem) {
  .body {
    font-size: 16px;
  }
}

@media (min-width: null) {
  .body {
    font-size: 18px;
  }
}

/*=============================================
=            Foundation - typeography         =
=============================================*/
/**
 * This file defines the actual colors that will be used for styling. They will default to the
 * we defined in the _variable.scss file. This is our default palette, and devs can override this
 * with their own variables.
 */
/*=============================================
 =            Body text sizes                  =
 =============================================*/
/*=============================================
 =            Heading text sizes               =
 =============================================*/
/*=============================================
 =            Head font weights                =
 =============================================*/
.ds-button__container {
  line-height: 1.5715;
  position: relative;
  display: inline-block;
  font-weight: 400;
  white-space: nowrap;
  text-align: center;
  background-image: none;
  border: 1px solid transparent;
  box-shadow: 0 2px #00000004;
  cursor: pointer;
  transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  height: 32px;
  padding: 4px 15px;
  font-size: 14px;
  border-radius: 2px;
  color: #000000d9;
  border-color: #d9d9d9;
  background: #fff;
}

.ds-button__container-primary {
  background: var(--ds-btn-primary-bg, #0f62fe);
  color: var(--ds-btn-primary-color, #fff);
  border: 1px solid var(--ds-btn-primary-bg, #0f62fe);
}

.ds-button__container-primary:hover {
  background: var(--ds-btn-primary-bg-hover, #78a9ff);
  color: var(--ds-btn-primary-color, #fff);
  border: 1px solid var(--ds-btn-primary-bg-hover, #78a9ff);
}

.ds-button__container-primary-tertiary {
  background: #fff;
  color: var(--ds-btn-primary-bg, #0f62fe);
  border: 1px solid var(--ds-btn-primary-bg, #0f62fe);
}

.ds-button__container-primary-tertiary:hover {
  background: var(--ds-btn-primary-bg, #0f62fe);
  color: var(--ds-btn-primary-color, #fff);
  border: 1px solid var(--ds-btn-primary-bg, #0f62fe);
}

.ds-button_container-ghost {
  background: #fff;
  color: var(--ds-btn-primary-bg, #0f62fe);
  border: none;
}

.ds-button_container-primary-ghost:hover {
  background: #bebfc4;
  color: var(--ds-btn-primary-bg, #0f62fe);
  border: 1px solid #bebfc4;
}

.ds-button_container-primary-disabled {
  background: #bebfc4;
  color: #a3a5ad;
  border: 1px solid #bebfc4;
  cursor: not-allowed;
}

.ds-button__container-success {
  background: var(--ds-btn-success-bg, #198038);
  color: var(--ds-btn-success-color, #fff);
  border: 1px solid var(--ds-btn-success-bg, #198038);
}

.ds-button__container-success:hover {
  background: var(--ds-btn-success-bg-hover, #42be65);
  color: var(--ds-btn-success-color, #fff);
  border: 1px solid var(--ds-btn-success-bg-hover, #42be65);
}

.ds-button__container-success-tertiary {
  background: #fff;
  color: var(--ds-btn-success-bg, #198038);
  border: 1px solid var(--ds-btn-success-bg, #198038);
}

.ds-button__container-success-tertiary:hover {
  background: var(--ds-btn-success-bg, #198038);
  color: var(--ds-btn-success-color, #fff);
  border: 1px solid var(--ds-btn-success-bg, #198038);
}

.ds-button_container-ghost {
  background: #fff;
  color: var(--ds-btn-success-bg, #198038);
  border: none;
}

.ds-button_container-success-ghost:hover {
  background: #bebfc4;
  color: var(--ds-btn-success-bg, #198038);
  border: 1px solid #bebfc4;
}

.ds-button_container-success-disabled {
  background: #bebfc4;
  color: #a3a5ad;
  border: 1px solid #bebfc4;
  cursor: not-allowed;
}

.ds-button__container-error {
  background: var(--ds-btn-error-bg, #da1e28);
  color: var(--ds-btn-error-color, #fff);
  border: 1px solid var(--ds-btn-error-bg, #da1e28);
}

.ds-button__container-error:hover {
  background: var(--ds-btn-error-bg-hover, #fa4d56);
  color: var(--ds-btn-error-color, #fff);
  border: 1px solid var(--ds-btn-error-bg-hover, #fa4d56);
}

.ds-button__container-error-tertiary {
  background: #fff;
  color: var(--ds-btn-error-bg, #da1e28);
  border: 1px solid var(--ds-btn-error-bg, #da1e28);
}

.ds-button__container-error-tertiary:hover {
  background: var(--ds-btn-error-bg, #da1e28);
  color: var(--ds-btn-error-color, #fff);
  border: 1px solid var(--ds-btn-error-bg, #da1e28);
}

.ds-button_container-ghost {
  background: #fff;
  color: var(--ds-btn-error-bg, #da1e28);
  border: none;
}

.ds-button_container-error-ghost:hover {
  background: #bebfc4;
  color: var(--ds-btn-error-bg, #da1e28);
  border: 1px solid #bebfc4;
}

.ds-button_container-error-disabled {
  background: #bebfc4;
  color: #a3a5ad;
  border: 1px solid #bebfc4;
  cursor: not-allowed;
}

.ds-button__container-warning {
  background: var(--ds-btn-warning-bg, #f1c21b);
  color: var(--ds-btn-warning-color, #fff);
  border: 1px solid var(--ds-btn-warning-bg, #f1c21b);
}

.ds-button__container-warning:hover {
  background: var(--ds-btn-warning-bg-hover, #f1c21b);
  color: var(--ds-btn-warning-color, #fff);
  border: 1px solid var(--ds-btn-warning-bg-hover, #f1c21b);
}

.ds-button__container-warning-tertiary {
  background: #fff;
  color: var(--ds-btn-warning-bg, #f1c21b);
  border: 1px solid var(--ds-btn-warning-bg, #f1c21b);
}

.ds-button__container-warning-tertiary:hover {
  background: var(--ds-btn-warning-bg, #f1c21b);
  color: var(--ds-btn-warning-color, #fff);
  border: 1px solid var(--ds-btn-warning-bg, #f1c21b);
}

.ds-button_container-ghost {
  background: #fff;
  color: var(--ds-btn-warning-bg, #f1c21b);
  border: none;
}

.ds-button_container-warning-ghost:hover {
  background: #bebfc4;
  color: var(--ds-btn-warning-bg, #f1c21b);
  border: 1px solid #bebfc4;
}

.ds-button_container-warning-disabled {
  background: #bebfc4;
  color: #a3a5ad;
  border: 1px solid #bebfc4;
  cursor: not-allowed;
}

.ds-button__container-sm {
  height: 1.5rem;
  padding: 0.5rem;
  font-size: 0.875rem;
  line-height: 0.5rem;
}

.ds-button__container-base {
  height: 2rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  line-height: 0.5rem;
}

.ds-button__container-lg {
  height: 2.5rem;
  padding: 1rem;
  font-size: 1rem;
  line-height: 0.5rem;
}
