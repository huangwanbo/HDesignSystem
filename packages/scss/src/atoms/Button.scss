@import "foundation/all";
$btn-size: (
  sm: 1.5rem,
  base: 2rem,
  lg: 2.5rem,
);
$btn-type: (
  primary: "primary",
  success: "success",
  error: "error",
  warning: "warning",
);
$btn-bg: (
  primary: $btn-primary-bg,
  success: $btn-success-bg,
  error: $btn-error-bg,
  warning: $btn-warning-bg,
);
$btn-bg-light: (
  primary: $btn-primary-bg-hover,
  success: $btn-success-bg-hover,
  error: $btn-error-bg-hover,
  warning: $btn-warning-bg-hover,
);
$btn-color: (
  primary: $btn-primary-color,
  success: $btn-success-color,
  error: $btn-error-color,
  warning: $btn-warning-color,
);
.ds-button__container {
  line-height: 1.5715;
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
  font-weight: 400;
  white-space: nowrap;
  text-align: center;
  background-image: none;
  border: 1px solid transparent;
  box-shadow: 0 2px #00000004;
  cursor: pointer;
  transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  user-select: none;
  touch-action: manipulation;
  height: 32px;
  padding: 4px 15px;
  font-size: 14px;
  border-radius: 2px;
  color: #000000d9;
  border-color: #d9d9d9;
  background: rgb(255, 253, 253);
}
@each $type, $value in $btn-type {
  .ds-button__container-#{$type} {
    background: map-get($btn-bg, $type);
    color: map-get($btn-color, $type);
    border: 1px solid map-get($btn-bg, $type);
  }
  .ds-button__container-#{$type}:hover {
    background: map-get($btn-bg-light, $type);
    color: map-get($btn-color, $type);
    border: 1px solid map-get($btn-bg-light, $type);
  }
  .ds-button__container-#{$type}-tertiary {
    background: $white;
    color: map-get($btn-bg, $type);
    border: 1px solid map-get($btn-bg, $type);
  }
  .ds-button__container-#{$type}-tertiary:hover {
    background: map-get($btn-bg, $type);
    color: map-get($btn-color, $type);
    border: 1px solid map-get($btn-bg, $type);
  }
  .ds-button__container-ghost {
    background: $white;
    color: map-get($btn-bg, $type);
    border: none;
  }
  .ds-button__container-#{$type}-ghost:hover {
    background: $gray-light;
    color: map-get($btn-bg, $type);
    border: 1px solid $gray-light;
  }
  .ds-button__container-#{$type}-disabled {
    background: map-get($btn-bg-light, $type);
    color: $white;
    border: 1px solid map-get($btn-bg-light, $type);
    cursor: not-allowed;
  }
}
@each $size, $value in $btn-size {
  .ds-button__container-#{$size} {
    height: $value;
    @if $size == sm {
      padding: map-get($spacing, "xxs");
      font-size: map-get($fontSize, "sm");
      line-height: 0.5rem;
    } @else if $size == base {
      padding: map-get($spacing, "xs");
      font-size: map-get($fontSize, "sm");
      line-height: 0.5rem;
    } @else {
      padding: map-get($spacing, "sm");
      font-size: map-get($fontSize, "base");
      line-height: 0.5rem;
    }
  }
}
