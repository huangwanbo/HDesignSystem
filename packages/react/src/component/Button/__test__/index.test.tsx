//@ts-nocheck
import React from "react";
import Button from "../Button";
import { shallow, mount, render } from "enzyme";
import { act } from "react-dom/test-utils";

describe("Button mount and unMount", () => {
  it("button could be updated an unmounted without errors", () => {
    const wrapper = mount(<Button />);
    expect(() => {
      wrapper.setProps({});
      wrapper.unmount();
    }).not.toThrow();
  });
});

describe("Button", () => {
  it("click callback correctly", () => {
    const mockFn = jest.fn();
    const component = mount(<Button onClick={mockFn} />);
    const button = component.find("button");
    button.simulate("click");
    const mockFnCallLength = mockFn.mock.calls.length;
    expect(mockFnCallLength).toBe(1);

    act(() => {
      component.setProps({
        disabled: true,
      });
    });

    button.simulate("click");
    expect(mockFn.mock.calls.length).toBe(mockFnCallLength);
  });

  it("render multiple children correctly", () => {
    const button = mount(<Button>1 2 3 </Button>);
    expect(button.text()).toEqual("1 2 3 ");
  });

  it("renders normal correctly", () => {
    const wrapper = shallow(<Button>按钮</Button>);
    expect(wrapper.find(".ds-button__container")).toExist();
    expect(wrapper.find(".ds-button__container-primary")).toExist();
    expect(wrapper.find(".ds-button__container-base")).toExist();
  });

  it("renders warming correctly", () => {
    const wrapper = shallow(<Button type="warning">按钮</Button>);
    expect(wrapper.find(".ds-button__container")).toExist();
    expect(wrapper.find(".ds-button__container-warning")).toExist();
    expect(wrapper.find(".ds-button__container-base")).toExist();
  });

  it("renders error correctly", () => {
    const wrapper = shallow(<Button type="error">按钮</Button>);
    expect(wrapper.find(".ds-button__container")).toExist();
    expect(wrapper.find(".ds-button__container-error")).toExist();
    expect(wrapper.find(".ds-button__container-base")).toExist();
  });

  it("renders success correctly", () => {
    const wrapper = shallow(<Button type="success">按钮</Button>);
    expect(wrapper.find(".ds-button__container")).toExist();
    expect(wrapper.find(".ds-button__container-success")).toExist();
    expect(wrapper.find(".ds-button__container-base")).toExist();
  });

  it("renders ghost correctly", () => {
    const wrapper = shallow(<Button type="ghost">按钮</Button>);
    expect(wrapper.find(".ds-button__container")).toExist();
    expect(wrapper.find(".ds-button__container-ghost")).toExist();
    expect(wrapper.find(".ds-button__container-base")).toExist();
  });

  it("renders size correctly", () => {
    const baseWrapper = shallow(<Button size="base">按钮</Button>);
    expect(baseWrapper.find(".ds-button__container-base")).toExist();
    const lgWrapper = shallow(<Button size="lg">按钮</Button>);
    expect(lgWrapper.find(".ds-button__container-lg")).toExist();
    const smWrapper = shallow(<Button size="sm">按钮</Button>);
    expect(smWrapper.find(".ds-button__container-sm")).toExist();
  });
});
