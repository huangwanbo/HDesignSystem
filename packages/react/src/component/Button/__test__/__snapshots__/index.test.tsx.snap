// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders Chinese characters correctly 1`] = `
<button
  class="ds-button__container ds-button__container-primary ds-button__container-base"
  type="button"
>
  按钮
</button>
`;

exports[`Button renders characters correctly 1`] = `
<button
  class="ds-button__container ds-button__container-primary ds-button__container-base"
  type="button"
>
  按钮
</button>
`;
